### PR TITLE
Material reflection

### PR DESCRIPTION
**Pull request by SamGondelman**
Using the new graphics scripting interface, you can query a shape entity, model entity, model overlay, or avatar for its materials and material names (if it has any).

This allows you to easily copy the materials from things into new material entities to apply them to other things.

Test plan:
- run [this](https://gist.githubusercontent.com/SamGondelman/f095df9fdc0a93272343374fc45850f6/raw/f66125e6fcb0605571bdd0ce212765026dbd9c6a/MaterialPicker.js) in an empty domain where you have edit rights.
- you should see two boxes and three models in front of you
- click the two boxes.  The middle mesh should change color to match the cube you click.
- when you click a box, an empty line should print in the log.
- click on various submeshes of the top right model (try: bow, shirt, arm).  The middle model should change appearance each time (the textures will be jumbled but that's fine).
- when you click on the top right model, this should print in the logs: `StingrayPBS3,StingrayPBS2,StingrayPBS1,StingrayPBS3,StingrayPBS1`
- click the bottom right model.  The middle model should return to normal.
- this should be printed in the logs: `StingrayPBS1`
- go into third person and click in the middle of your avatar.  If you hit it (avatar raypicks collide against the T-pose, not the animated mesh), the middle model should change appearance (what it looks like will depend on what avatar you're wearing and which submesh you select)

### CODE REVIEW

File: libraries/gpu/src/gpu/Texture.h [hunk 1]
```diff
@@ -647,13 +647,11 @@ typedef std::vector<TextureView> TextureViews;
 // It provides the mechanism to create a texture using a customizable TextureLoader
 class TextureSource {
 public:
-    TextureSource();
-    ~TextureSource();
+    TextureSource(const QUrl& url, int type = 0) : _imageUrl(url), _type(type) {}
```
Comment (SamGondelman) [inline libraries/gpu/src/gpu/Texture.h]: image depends on gpu, so gpu can't depend on image to get image::TextureUsage::Type here, so I stored it as an int.  Is there a better way?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7574/app_12515_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7574/HighFidelity-Beta-PR12515-a4b04d9d4391e5313945aa91ec655829a9df3229.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7574/HighFidelity-Beta-PR12515-a4b04d9d4391e5313945aa91ec655829a9df3229.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by qlimnavarro** [comment]
Fail, empty line does not print when the cube is selected. In HMD, user can only far grab cubes with hand controllers and can only select then with the mouse. Video and Log attached.
[PR12515 Files.zip](https://github.com/highfidelity/hifi/files/1767749/PR12515.Files.zip)
@conklin94122

**Comment by SamGondelman** [comment]
@conklin94122 @qlimnavarro whoops sorry I should have been clearer, the script only works with the system mouse, not hand controllers.

**Comment by conklin94122** [comment]
@SamGondelman To clarify, are you saying that our testing it works as intended and should PASS, or are you agreeing that it's a fail and giving further instruction on how to perform later testing?

**Comment by conklin94122** [comment]
QA, please retest taking Sam's updated info into account.

**Comment by qlimnavarro** [comment]
Currently Testing.

**Comment by qlimnavarro** [comment]
Fail, step 4: No empty line in the log when selecting the boxes. Instead this appears: 

[02/28 12:49:31] [DEBUG] [hifi.ui] setKeyboardRaised:  Desktop_QMLTYPE_208(0x19cdf0eb700, name = "desktop") , raised:  false , numeric:  false , password:  false
[02/28 12:49:31] [DEBUG] [hifi.ui] setKeyboardRaised:  QQuickRootItem(0x19caed5bcd0) , raised:  false , numeric:  false , password:  false
[02/28 12:49:31] [DEBUG] [hifi.ui] setKeyboardRaised:  Desktop_QMLTYPE_208(0x19cdf0eb700, name = "desktop") , raised:  false , numeric:  false , password:  false
[02/28 12:49:31] [DEBUG] [default] GeometryCache::getMeshProxyListFromShape 0x19caad90de8 "Cube"
[02/28 12:49:31] [DEBUG] [hifi.scriptengine.script] [MaterialPicker.js]

All other test steps pass.
@conklin94122 
[hifi-log_2018-02-28_12.46.48_c0210289-46bc-4982-95fe-cb6bb7b4818c.txt](https://github.com/highfidelity/hifi/files/1768601/hifi-log_2018-02-28_12.46.48_c0210289-46bc-4982-95fe-cb6bb7b4818c.txt)
[hifi-log.txt](https://github.com/highfidelity/hifi/files/1768602/hifi-log.txt)

**Comment by SamGondelman** [comment]
Ahhh sorry again, `[02/28 12:49:31] [DEBUG] [hifi.scriptengine.script] [MaterialPicker.js]` is what I meant by empty line.  Sounds like a pass then, thank you!

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7651/app_12515_PR.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7651/HighFidelity-Beta-PR12515-a66125fbb781af8faf8fa1cef32628974291e5b0.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7651/HighFidelity-Beta-PR12515-a66125fbb781af8faf8fa1cef32628974291e5b0.dmg

**Comment by qlimnavarro** [comment]
Pass - @conklin94122

**Comment by qlimnavarro** [comment]
Currently Testing.

**Comment by qlimnavarro** [comment]
Pass @conklin94122

**Comment by conklin94122** [comment]
@SamGondelman Needs dev review

**Comment by misslivirose** [comment]
We've been adding JSDoc style comments to anything that can be accessed from the JS scripting interfaces - the structure of this looks good to me from the scripting side, but it should be added to the documentation and surfaced as an available SDK for scripters to use.

**Comment by samcake** [comment]
@samuel, @misslivirose is asking for documentation

**Comment by SamGondelman** [comment]
@misslivirose docs added (they point to one thing that is in @ctrlaltdavid's Entities API work, so that one link won't work until https://github.com/highfidelity/hifi/pull/12526 is merged)

gonna put QA approved back because that was just a jsdoc change

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/7839/app_12515_PR.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/7839/HighFidelity-Beta-PR12515-0aacdcd5580f4d63530023d694d2c9a8ec5c72a0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/7839/HighFidelity-Beta-PR12515-0aacdcd5580f4d63530023d694d2c9a8ec5c72a0.exe

**Pull request merged**